contracts/DefaultStrategy.sol

```diff
@@ -522,248 +436,288 @@ contract DefaultStrategy is UUPSOwnableUpgradeable, UsingRegistryUpgradeable, Ma
-                IManager(manager).toStakedCelo(withdrawalsPerGroup[i])
+            (address lesserKey, address greaterKey) = getLesserAndGreaterOfActiveGroups(
+                votedGroup,
+                stCELOInGroup[votedGroup],
+                withdraw
             );
-            if (availableVotes % numberGroupsWithdrawn > i) {
-                withdrawalsPerGroup[i]--;
+            if ((lesserKey != greaterKey || activeGroups.getNumElements() == 1) && sorted) {
+                activeGroups.update(votedGroup, stCELOInGroup[votedGroup], lesserKey, greaterKey);
+                votedGroup = withdraw ? activeGroups.getHead() : activeGroups.getTail();
+            } else {
+                unsortedGroups.add(votedGroup);
+                sorted = false;
+                if (withdraw) {
+                    (, votedGroup, ) = activeGroups.get(votedGroup);
+                } else {
+                    (, , votedGroup) = activeGroups.get(votedGroup);
+                }
             }
         }
 
-        return (groupsWithdrawn, withdrawalsPerGroup);
-    }
+        if (votesAmount != 0) {
+            revert NotAbleToDistributeVotes();
+        }
 
-    /**
-     * @notice Adds value to totals of group strategy and
-     * total stCELO in all group strategies.
-     * @param strategy The validator group that we are adding to.
-     * @param stCeloAmount The added amount of stCELO.
-     */
-    function addToStrategyTotalStCeloVotes(address strategy, uint256 stCeloAmount) internal {
-        defaultStrategyTotalStCeloVotes[strategy] += stCeloAmount;
-        totalStCeloInDefaultStrategy += stCeloAmount;
-    }
+        finalGroups = new address[](groupsIndex);
+        finalVotes = new uint256[](groupsIndex);
 
-    /**
-     * @notice Subtracts value from totals of group strategy and
-     * total stCELO in all group strategies.
-     * @param strategy The validator group that we are adding to.
-     * @param stCeloAmount The subtracted amount of stCELO.
-     */
-    function subtractFromStrategyTotalStCeloVotes(address strategy, uint256 stCeloAmount) internal {
-        defaultStrategyTotalStCeloVotes[strategy] -= stCeloAmount;
-        totalStCeloInDefaultStrategy -= stCeloAmount;
+        for (uint256 j = 0; j < groupsIndex; j++) {
+            finalGroups[j] = groups[j];
+            finalVotes[j] = votes[j];
+        }
     }
 
-    /**
-     * @notice Returns a list of group addresses with their corresponding
-     * current total votes, sorted by the number of votes, and the total number
-     * of votes in the system.
-     * @param groups The array of addresses of the groups to sort.
-     * @return The array of GroupWithVotes structs, sorted by number of votes.
-     * @return The total number of votes assigned to active groups.
-     */
-    function getSortedGroupsWithVotes(address[] memory groups)
-        internal
+    function getLesserAndGreaterOfActiveGroupsDeposit(address originalKey, uint256 newValue)
```

Comment:
One very fancy thing that could potentially be done to deduplicate these two functions would be to use function references (https://docs.soliditylang.org/en/v0.8.11/types.html#function-types), define something like `iterateToNext`, `iterateToPrevious`, then select one depending on whether we want to move up or down the list.